LDAP search member of filter

BioinformaticsPlatformWIV-ISP opened the pull request:
Hello,

Latest version of the pr #8628 (previously #7543) generated from a clone of the origin galaxy repo dev branch. This should avoid merge conflics and is up to date with the dev branch.

Commit:
LDAP MemberOf filter

Commit:
remove additional empty line

nsoranzo committed:
Re-use `transform_publicname()` function

nsoranzo commented:
@BioinformaticsPlatformWIV-ISP Can you please check the commit I've pushed to this PR? The rest looks good to me!

BioinformaticsPlatformWIV-ISP commented on lib/galaxy/security/validate_user_input.py:
```diff
@@ -65,24 +67,27 @@ def validate_publicname(trans, publicname, user=None):
-def transform_publicname(trans, publicname, user=None):
-    # User names must be at least four characters in length and contain only lower-case
-    # letters, numbers, and the '-' character.
+def transform_publicname(trans, publicname, user=None, check_dup=True):
```
`transform_publicname` has two mandatory argument `trans` and `publicname`.  Only the first one is receive when called from `authenticate`. Lead to crash (exception raise) in this case. In addition, `trans` is only used when duplication check is enabled.

BioinformaticsPlatformWIV-ISP commented on lib/galaxy/security/validate_user_input.py:
```diff
@@ -65,24 +67,27 @@ def validate_publicname(trans, publicname, user=None):
+    if publicname not in ['None', None, '']:
         publicname = publicname.lower()
         publicname = re.sub(VALID_PUBLICNAME_SUB, FILL_CHAR, publicname)
         publicname = publicname.ljust(PUBLICNAME_MIN_LEN + 1, FILL_CHAR)[:PUBLICNAME_MAX_LEN]
```
`FILL_CHAR` is set to '-' (minus) whereas our old clean function was using '_ ' (underscore) instead. This is valid but a little annoying for us. We will have two kind of pattern for galaxy public username. One with '_' and the other one with '-' depending of when the username has been generated (before or after galaxy 20.1) .  I think that as admin we can deal with it but for ours user it will be awkward.

nsoranzo replied:
I guess you can use a database query to update the existing Galaxy usernames to use `-`?

BioinformaticsPlatformWIV-ISP commented on lib/galaxy/security/validate_user_input.py:
```diff
@@ -65,24 +67,27 @@ def validate_publicname(trans, publicname, user=None):
+    FILL_CHAR is used to extend or replace characters.
+    """
     # TODO: Enhance to allow generation of semi-random publicnnames e.g., when valid but taken
     if user and user.username == publicname:
```
Do you have any use case in mind where `transform_publicname` is called with an existing user and a matching publicname ?

nsoranzo replied:
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/tool_shed/controllers/user.py#L157

nsoranzo committed:
Remove unused parameter

BioinformaticsPlatformWIV-ISP commented on lib/galaxy/auth/providers/ldap_ad.py:
```diff
@@ -204,12 +209,18 @@ def authenticate(self, email, username, password, options):
         return (True,
                 _get_subs(options, 'auto-register-email', params),
-                _get_subs(options, 'auto-register-username', params),
+                transform_publicname(_get_subs(options, 'auto-register-username', params), check_dup=False),
```
See the other comment. Crash when called this way, without the `trans` parameter.  If there is no access to `trans ` in this context, maybe separate the public name transformation and the duplication check in two functions and only call the transformation specific one here ?

nsoranzo replied:
Yes, sorry for the trouble, that's a good plan, I'll work on a patch.

nsoranzo replied:
It should be fixed now.

nsoranzo committed:
Remove unnecessary call to `transform_publicname()`

The username is already validated both client-side (`validateString()`
JavaScript function in `templates/webapps/tool_shed/user/register.mako`) and
server-side (call to `__validate()` method in
`lib/galaxy/webapps/tool_shed/controllers/user.py`).

Simplify `transform_publicname()` even more as a consequence.

BioinformaticsPlatformWIV-ISP commented:
Your latest changes have fix the 'transform_publicname' crash, authentication process is now back to normal. Thank you for your help.
